### ISSUE TITLE
New wraparound mode that never switches rows automatically?

### ISSUE THREAD

**Issue by mikehardy** [mzur/gnome-shell-wsmatrix#95]
Hi there! First, it's important to say I love this extension, thank you

Second, this is a feature request so it may have already been considered and discarded, but I looked and didn't see one. Either way, it's possible there is no interest to implement it so I'd have to implement it myself, that's all fine :-)

Here is what I am personally missing though - I want a wraparound mode that never switches rows unless I explicitly switch by asking to move up or down.

Concretely, let's say I have a 2x2 matrix, screens 1, 2, 3, 4 like the terrible ASCII art below:

```
----------
| 1 |  2 |
-----------
| 3 | 4  |
----------
```

If I am on screen 1, and hit "move right one desktop", it should go to 2, and if I do it again I want it to go to 1. Staying on the row.

In my experience it always goes to 3. I only want it to go to 3 if I am on 1 and I say "move down one desktop"

Is this already possible? If not, could I add a wraparound mode?

Cheers

**Comment by ebeem** [comment]
can you try the wraparound mode `None`?
I think this is what you're looking for, try it out and let us know.

There are 3 different wraparound modes in the preferences
```
None
Next/Previous
Rows/Columns
```

**Comment by mzur** [comment]
I think the `Rows/Columns` mode should do what you want.

**Comment by ebeem** [comment]
sorry, I didn't understand your requirements well.
I think @mzur suggestion about `Rows/Columns` is actually what you are looking for
this way you will basically be looping around the row or column

assuming you have a 3x3 matrix, and you're on workspace#3
```
1	2     [3]
4	5	6
7	8	9
```
the behavior when you move right
- `None mode`: nothing happens
- `Rows/Columns`: go to workspace#1 (loops the row)
- `Next/Previous`: go to workspace#4 (next row)

the behavior when you move up
- `None mode`: nothing happens
- `Rows/Columns`: go to workspace#9 (loops the column)
- `Next/Previous`: go to workspace#8 (previous column)


now if you have a 2x2 setup with `Rows/Columns` mode
and you go right and the workspace changes to `workspace#3` instead of `workspace#1`, then this is not the expected behavior, and there's something wrong with the extension in your setup.

**Comment by mikehardy** [comment]
> now if you have a 2x2 setup with `Rows/Columns` mode and you're on `workspace#2`
> and you go right and the workspace changes to `workspace#3` instead of `workspace#1`, then this is not the expected behavior, and there's something wrong with the extension in your setup.

This! I am using a 2-row/6-column in practice. I have Rows/Columns. I am on 6 (top right) and I say "move right", it goes to 7 (bottom left) instead of 1 (top right).

So how to diagnose that?

I was just prompted that there was an update, I'm on Ubuntu 19.10. 

I scanned the output of `journalctl /usr/bin/gnome-shell` and frankly it's got way too much personal information in it to consider attaching. There must be a sequence of commands that could help triangulate that doesn't involve personally identifying information?

Perhaps this snippet grepped out from `apt list`:
gnome-shell/eoan-updates,now 3.34.1+git20191024-1ubuntu1~19.10.1 amd64 [installed]

and

`cat $HOME/.local/share/gnome-shell/extensions/wsmatrix@martin.zurowietz.de/metadata.json `
```json
{
  "_generated": "Generated by SweetTooth, do not edit",
  "description": "Arrange workspaces in a two dimensional grid with workspace thumbnails.",
  "gettext-domain": "wsmatrix",
  "keybindings-schema": "org.gnome.shell.extensions.wsmatrix-keybindings",
  "name": "Workspace Matrix",
  "settings-schema": "org.gnome.shell.extensions.wsmatrix-settings",
  "shell-version": [
    "3.34"
  ],
  "url": "https://github.com/mzur/gnome-shell-wsmatrix",
  "uuid": "wsmatrix@martin.zurowietz.de",
  "version": 17
```

What could be wrong? Should I just try hacking around in there and reloading gnome-shell to add logging and try things?

**Comment by mikehardy** [comment]
you guys have been so responsive, just a note to say I'm heading to remote islands for 10 days so no rush, and sorry I won't add anything, if more info is needed, or try anything for a little bit in the absence of other prompts. Cheers

**Comment by ebeem** [comment]
@mikehardy I have just tried it on a `2-row/6-column` setup
it's working well, when I am top right (`workspace#6`) and I go right, it goes to `workspace#1` if I am in `rows/columns` mode, it goes `workspace#7` if I am in `next/previous` mode

**Comment by mikehardy** [comment]
Fascinating. I'll dig in to my local installation and settings then - thanks for attempting to reproduce

**Comment by mzur** [comment]
I tested it as well with two rows and six columns in my Ubuntu 19.10 VM with the same GNOME version and it works as expected. @mikehardy What other extensions do you have installed?

Just out of curiosity: What do you consider personal information in `journalctl /usr/bin/gnome-shell` besides your username and the extensions you have installed? I could add recommendations on how to anonymize it before posting it here.

**Comment by mikehardy** [comment]
I have quite a few and interactions are a strong possibility. I'll dig in

My calendar events with info are dumped into journalctl, no idea why but can't post that 😁

**Comment by ebeem** [comment]
any news on this?
@mikehardy are you still facing the issue? can you let us provide us with a list of gnome extensions installed on your environement?

**Comment by mikehardy** [comment]
Sorry I did not report back, I did not have time to fully investigate and I didn't want to waste others' time further when it appeared the resolution was going to be that it was a local problem. And now after upgrading from Ubuntu 19 to 20 I can only say that the problem no longer occurs, and I have no idea why really. The worst sort of resolution in that it provides no new knowledge but, one less bug at least

If there is any value I can provide then it is only in saying this extension is amazing, I would be very sad if I didn't have it - thanks for the work on it. Cheers
